File: SDK-sources/OutbrainSDK/SFWidget/SFWidgetMessageHandler.m
//
//  SFWidgetMessageHandler.m
//  OutbrainSDK
//
//  Created by Oded Regev on 27/06/2021.
//  Copyright © 2021 Outbrain. All rights reserved.
//

#import "SFWidgetMessageHandler.h"
#import "SFWidget.h"
#import "OBErrorReporting.h"

@implementation SFWidgetMessageHandler

- (void)userContentController:(WKUserContentController *)userContentController didReceiveScriptMessage:(WKScriptMessage *)message {
    @try  {
        if (! [message.name isEqualToString:@"ReactNativeWebView"]) {
            NSLog(@"SFWidgetMessageHandler - message is not ReactNativeWebView");
            return;
        }
        NSString *jsonString = message.body;
        NSData *data = [jsonString dataUsingEncoding:NSUTF8StringEncoding];
        NSDictionary *msgBody = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
        
        
        if ([msgBody valueForKey:@"height"]) {
            NSInteger newHeight = [[msgBody valueForKey:@"height"] integerValue];
            [self.delegate didHeightChanged:newHeight];
        }
        if ([msgBody valueForKey:@"bridgeParams"]) {
            NSString *bridgeParam = [msgBody valueForKey:@"bridgeParams"];
            NSLog(@"SFWidgetMessageHandler received bridgeParam: %@", bridgeParam);
            [[NSNotificationCenter defaultCenter] postNotificationName:SFWIDGET_BRIDGE_PARAMS_NOTIFICATION object:self userInfo:@{@"bridgeParams" : bridgeParam}];
        }
        if ([msgBody valueForKey:@"t"]) {
            NSString *tParam = [msgBody valueForKey:@"t"];
            NSLog(@"SFWidgetMessageHandler received t param: %@", tParam);
        }
        if ([msgBody valueForKey:@"t"]) {
            NSString *tParam = [msgBody valueForKey:@"t"];
            NSLog(@"SFWidgetMessageHandler received t param: %@", tParam);
            [[NSNotificationCenter defaultCenter] postNotificationName:SFWIDGET_T_PARAM_NOTIFICATION object:self userInfo:@{@"t" : tParam}];
        }
        if ([msgBody valueForKey:@"url"]) {
            NSString *urlString = [msgBody valueForKey:@"url"];
            NSString *type = [msgBody valueForKey:@"type"];
            if ([type isEqualToString:@"organic-rec"]) {
                NSString *orgUrl = [msgBody valueForKey:@"orgUrl"];
                [self.delegate didClickOnOrganicRec:urlString orgUrl:orgUrl];
            }
            else {
                [self.delegate didClickOnRec:urlString];
            }
        }
        if ([msgBody valueForKey:@"event"] && ([[msgBody valueForKey:@"event"] isKindOfClass:[NSDictionary class]])) {
            NSMutableDictionary *eventData = [[msgBody valueForKey:@"event"] mutableCopy];
            NSString *eventName = @"";
            if ([[eventData valueForKey:@"name"] isKindOfClass:[NSString class]]) {
                eventName = [eventData valueForKey:@"name"];
            }
            else {
                eventName = @"** event_name_missing **";
            }
            [eventData removeObjectForKey:@"name"];
            
            if ([self verifyAllKeysAreTypeString:eventData]) {
                [self.delegate widgetEvent:eventName additionalData:eventData];
            }
        }
        if ([msgBody valueForKey:@"errorMsg"]) {
            NSString *errorMsg = [msgBody valueForKey:@"errorMsg"];
            errorMsg = [NSString stringWithFormat:@"Bridge: %@", errorMsg];
            [[OBErrorReporting sharedInstance] reportErrorToServer:errorMsg];
        }
        
    } @catch (NSException *exception) {
        NSLog(@"Exception in SFWidgetMessageHandler() - %@",exception.name);
        NSLog(@"Reason: %@ ",exception.reason);
        NSString *errorMsg = [NSString stringWithFormat:@"Exception in SFWidgetMessageHandler - %@ - reason: %@", exception.name, exception.reason];
        [[OBErrorReporting sharedInstance] reportErrorToServer:errorMsg];
    }
}

- (BOOL) verifyAllKeysAreTypeString:(NSDictionary *) dict {
    NSEnumerator *keyEnumerator = [dict keyEnumerator];
    id key;
    while ((key = [keyEnumerator nextObject])) {
        if (![key isKindOfClass:[NSString class]]) {
            NSLog(@"OBSDK The key (%@) does not belong to NSString class", key);
            return NO;
        }
    }
    return YES;
}

@end

File: SwiftSDK-sources/OutbrainSDK/SFWidget/SFWidgetMessageHandler.swift
//
//  SFWidgetMessageHandler.swift
//  OutbrainSDK
//
//  Created by Shai Azulay on 25/07/2023.
//

import Foundation
import WebKit

class SFWidgetMessageHandler: NSObject, WKScriptMessageHandler {
    
    private let messageHandler = "message-handler"
    weak var delegate: SFWidget?
    
    
    func userContentController(
        _ userContentController: WKUserContentController,
        didReceive message: WKScriptMessage
    ) {
        guard message.name == "ReactNativeWebView" else {
            Outbrain.logger.debug(
                "SFWidgetMessageHandler - message is not ReactNativeWebView",
                domain: messageHandler
            )
            return
        }
        
        guard let jsonString = message.body as? String else {
            Outbrain.logger.debug(
                "SFWidgetMessageHandler - Invalid message body",
                domain: messageHandler
            )
            return
        }
        
        guard let data = jsonString.data(using: .utf8) else {
            Outbrain.logger.debug(
                "SFWidgetMessageHandler - Error converting message body to data",
                domain: messageHandler
            )
            return
        }
            
        do {
            let msgBody = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] ?? [:]
            
            callDelegateOnChangeHeight(msgBody)
            handleBridgeParams(msgBody)
            handleTParam(msgBody)
            handleClickMessage(msgBody)
            callDelegateOnValidEvents(msgBody)
            handleErrorMessage(msgBody)
            handleSettingsMessage(msgBody)
        } catch {
            let errorMsg = "Exception in SFWidgetMessageHandler - \(error)"
            Outbrain.logger.error("SFWidgetMessageHandler - Error converting message body to data", domain: messageHandler)
            Outbrain.logger.error(errorMsg, domain: messageHandler)
            delegate?.errorReporter?.setMessage(message: errorMsg).reportErrorToServer()
        }
    }
    
    
    private func handleBridgeParams(_ msg: [String: Any]) {
        guard let bridgeParam = msg["bridgeParams"] as? String else { return }
        
        Outbrain.logger.debug(
            "SFWidgetMessageHandler received bridgeParam: \(bridgeParam)",
            domain: messageHandler
        )
        
        NotificationCenter.default.post(
            name: NSNotification.Name(rawValue: SFConsts.SFWIDGET_BRIDGE_PARAMS_NOTIFICATION),
            object: self,
            userInfo: ["bridgeParams": bridgeParam]
        )
    }
    
    
    private func handleTParam(_ msg: [String: Any]) {
        guard let tParam = msg["t"] as? String else { return }
        
        Outbrain.logger.debug(
            "SFWidgetMessageHandler received t param: \(tParam) - Not posting notification (using bridgeParams instead)",
            domain: messageHandler
        )
    }
    
    
    private func handleClickMessage(_ msg: [String: Any]) {
        guard let urlString = msg["url"] as? String,
              BridgeUtils.isValidURL(urlString) else {
            return
        }
        
        if let type = msg["type"] as? String, type == "organic-rec",
           let orgUrl = msg["orgUrl"] as? String {
            delegate?.didClickOnOrganicRec(urlString, orgUrl: orgUrl)
            return
        }
        
        delegate?.onRecClick(URL(string: urlString)!)
    }
    
    
    private func handleErrorMessage(_ msg: [String: Any]) {
        guard let errorMsg = msg["errorMsg"] as? String else { return }
        
        let errorMsgWithPrefix = "Bridge: \(errorMsg)"
        delegate?.errorReporter?.setMessage(message: errorMsgWithPrefix).reportErrorToServer()
    }
    
    
    private func callDelegateOnChangeHeight(_ msg: [String: Any]) {
        guard let height = msg["height"] as? CGFloat, msg["sender"] as? String == "resize" else { return }
        delegate?.messageHeightChange(height)
    }
    
    
    private func callDelegateOnValidEvents(_ msg: [String: Any]?) {
        guard var eventDict = msg?["event"] as? [String: Any] else { return }
        
        if let eventName = eventDict.removeValue(forKey: "name") as? String,
           eventDict.keys.compactMap({ $0 }).count == eventDict.keys.count {
            delegate?.widgetEvent(
                eventName: eventName,
                additionalData: eventDict
            )
        }
    }
    
    
    private func handleSettingsMessage(_ msg: [String: Any]) {
        guard let settings = msg["settings"] as? [String: Any] else { return }
        
        delegate?.onSettingsReceived(settings)
    }
}

